### PR TITLE
Activity indicator: system start on kubernetes

### LINKED ISSUE #3212: [mesheryctl] Activity Indicator: system start on Kubernetes
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior
```
➜  ~ mesheryctl system start
downloading manifest files from https://github.com/layer5io/meshery/tree/v0.5.23/install/deployment_yamls/k8s
Starting Meshery...
```

Users may sit staring at an inactive terminal for some undetermined amount of time, while Meshery starts on a Kubernetes system. Users deploying Meshery to Docker have status indicators as to provisioning activity, however, K8s users do not.

#### Desired Behavior
An activity spinner can be quickly incorporated into this command to show the user that the system is busy working on their request.

#### Implementation

@navendu-pottekkat has recently used an activity spinner library. This same library could be used here.

---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/layer5io/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by NagaHarshita**
Signed-off-by: NagaHarshita <harshitamarupaka@gmail.com>

**Description**

This PR fixes #3212. 
Added activity spinner to the system start on kubernetes. 
```
downloading manifest files from https://github.com/layer5io/meshery/tree/v0.5.25/install/deployment_yamls/k8s
Starting Meshery...
⣷ Deploying Meshery on Kubernetes.
```
After successful deployment.
```
downloading manifest files from https://github.com/layer5io/meshery/tree/v0.5.25/install/deployment_yamls/k8s
Starting Meshery...
...Meshery deployed on Kubernetes.
Meshery is started.
```
**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by leecalcote** [comment]
Hey! This is pretty neat!

File: mesheryctl/internal/cli/root/system/start.go [hunk 1]
```diff
@@ -341,14 +341,24 @@ func start() error {
...
 		log.Info("Starting Meshery...")
 
-		// apply the adapters mentioned in the config.yaml file to the Kubernetes cluster
-		err = utils.ApplyManifestFiles(manifests, RequestedAdapters, kubeClient, false, false)
+		spinner := utils.CreateDefaultSpinner("Deploying Meshery on Kubernetes.", "...Meshery deployed on Kubernetes.")
+		spinner.Start()
 
-		if err != nil {
-			return err
+		deadline := time.Now().Add(20 * time.Second)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/start.go:347]: If the deployment happens faster than 20 seconds, the user will have to wait anyway, correct? Why not have a lower deadline or do away with the deadline altogether?

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/start.go:347]: @nagaharshita, What do you think about this? How do we optimize for the least amount of unnecessary wait time?

Comment (NagaHarshita) [reply mesheryctl/internal/cli/root/system/start.go:347]: How about we remove the deadline so that the spinner runs for the exact time till the Meshery is deployed?

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/start.go:347]: Yes, let the spinner run until Meshery is deployed. It should not be an issue, if any error occurs inbetween, it will stop the execution and show the error.

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/start.go:347]: @leecalcote It is actually the other way around. The maximum time we wait is 20 seconds. If we deploy faster than that, then the spinner stops. That is the way the deadline works (also how it is implemented in `system stop`), will have to check this PR out.

File: mesheryctl/internal/cli/root/system/start.go [hunk 2]
```diff
@@ -341,14 +341,24 @@ func start() error {
...
+
+		for !(time.Now().After(deadline)) {
+			// apply the adapters mentioned in the config.yaml file to the Kubernetes cluster
+			err := utils.ApplyManifestFiles(manifests, RequestedAdapters, kubeClient, false, false)
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/start.go:351]: `ApplyManifestFiles` will just apply the manifests. It does not actually check if the pods are running. For that we need to check the status of the pods in `meshery` namespace.

**Review by navendu-pottekkat — CHANGES_REQUESTED** [review]
@NagaHarshita Could you check the recent, similar changes to `stop.go`? This is related to the comment below.

File: mesheryctl/internal/cli/root/system/start.go (hunk 1 above)
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/start.go:347]: Oh, I see. On a related note, we are allowing the user’s browser to open before Meshery Server is serving its UI. The check for this doesn’t seem to be working a new longer on either Docker or Kubernetes deployments.

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/start.go:347]: @metonymic-smokey ^^

File: mesheryctl/internal/cli/root/system/start.go [hunk 3]
```diff
@@ -341,14 +341,33 @@ func start() error {
 		// downloaded required files successfully now apply the manifest files
 		log.Info("Starting Meshery...")
 
+		spinner := utils.CreateDefaultSpinner("Deploying Meshery on Kubernetes.", "...Meshery deployed on Kubernetes.")
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/start.go:344]:
```suggestion
		spinner := utils.CreateDefaultSpinner("Deploying Meshery on Kubernetes", "Meshery deployed on Kubernetes.")
```

File: mesheryctl/internal/cli/root/system/start.go [hunk 4]
```diff
@@ -341,14 +341,33 @@ func start() error {
...
 		// apply the adapters mentioned in the config.yaml file to the Kubernetes cluster
 		err = utils.ApplyManifestFiles(manifests, RequestedAdapters, kubeClient, false, false)
-
 		if err != nil {
-			return err
+			break
 		}
 
-		log.Info("...Meshery deployed on Kubernetes.")
+		deadline := time.Now().Add(10 * time.Second)
```
Comment (navendu-pottekkat) [inline mesheryctl/internal/cli/root/system/start.go:353]: It might take longer than 10 sec on first install. @leecalcote @hexxdump what should we do here?

File: mesheryctl/internal/cli/root/system/start.go (hunk 3 above)
Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/start.go:344]: I think this should be better...

**Review by navendu-pottekkat — COMMENTED** [review]
Thank you @NagaHarshita . I have made some comments that we need to check.

File: mesheryctl/internal/cli/root/system/start.go (hunk 4 above)
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/start.go:353]: What’s the behavior after it hits the deadline? Does it back off and check every three seconds?

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/system/start.go:353]: No, it will basically skip the checking and proceed to say "Meshery is started."

We can let it check indefinitely but if there are any issues with the deployment then it will keep checking. Hence we have a timeout here. What should be the ideal "time out" time is the question here.

Comment (hexxdump) [reply mesheryctl/internal/cli/root/system/start.go:353]: As @leecalcote mentioned, we can not wait forever, there has to be an ultimate deadline.
Here are my thoughts: After the ultimate deadline is met, based on return value of utils.IsMesheryRunning(), we modify the FinalMsg to either “All pods started” or “Pod creation in progress, check kubectl get pods -n meshery before using Meshery”. I think we could modify IsMesheryRunning() or write a new utility to report the status of Pod/Deployment, rather than just checking for whether the deployment is listed or not(it could very well be a deployment in progress). Also, it would help, if we could actually display the status of deployment on the ultimate deadline.

**Comment by leecalcote** [comment]
@NagaHarshita let us (@hexxdump) know if a brainstorming session on the approach here would be helpful.

File: mesheryctl/internal/cli/root/system/start.go [hunk 5]
```diff
@@ -350,24 +350,30 @@ func start() error {
...
 				time.Sleep(1 * time.Second)
 			}
 		}
 
 		spinner.Stop()
 
-		log.Info("Meshery is started.")
+		podsStatus, err := utils.AreAllPodsRunning()
+		if !podsStatus {
+			log.Info("\nCurrently all the Meshery pods are not Running.\n Please check the status of the pods by executing “kubectl get pods -—namespace=meshery” before using meshery")
```
Comment (hexxdump) [inline mesheryctl/internal/cli/root/system/start.go:373]: Can we reword it to convey that the pods are being brought up?

Some Meshery pods have not come up yet. \n Please check the status of the pods by executing “kubectl get pods -—namespace=meshery” before using meshery".

File: mesheryctl/pkg/utils/healthcheck.go [hunk 6]
```diff
@@ -157,3 +157,29 @@ func IsMesheryRunning(currPlatform string) (bool, error) {
 
 	return false, nil
 }
+
+// AreAllPodsRunning checks if all the deployment pods under kubernetes are running
```
Comment (hexxdump) [inline mesheryctl/pkg/utils/healthcheck.go:161]: Looks good.

**Review by hexxdump — APPROVED** [review]

**Review by navendu-pottekkat — APPROVED** [review]
Looks good! Thank you @NagaHarshita 🚀

**Review by hexxdump — DISMISSED** [review]

**Comment by hexxdump** [comment]
@NagaHarshita, you need to sign off the commit I guess. Can you pls do that by following the steps recommended at:
https://github.com/meshery/meshery/pull/3256/checks?check_run_id=3002748430

**Comment by leecalcote** [comment]
@NagaHarshita you know that you're working on meaningful things when it take this long and this many reviews to land the PR. 😄 Good work!

**Review by leecalcote — APPROVED** [review]

**Comment by hexxdump** [comment]
> @NagaHarshita you know that you're working on meaningful things when it take this long and this many reviews to land the PR. 😄 Good work!

+1 to that. Good work @NagaHarshita !
